### PR TITLE
TinyLlama Processing Sink

### PR DESCRIPTION
**Pull request by destrex271**
Uses Llama3 model(from ollama) to derive insights on metrics for a given database and stores the results in postgres

References:
 - #32

### CODE REVIEW

**Comment by destrex271** [comment]
Model Updated to tinyllama for ease of user

**Comment by pashagolub** [comment]
Looks cool! Do you have any demo output? I'm really eager to see how it works :)

**Comment by destrex271** [comment]
Yep this is how we get the out put for some dummy metrics:

*Metrics*(We are picking the 10 most recent measurements for now so that my laptop does not blow up :)): 
```
       created_time        |                        data                         | metric_name | database_id | id | dbname 
----------------------------+-----------------------------------------------------+-------------+-------------+----+--------
 2024-09-18 15:09:17.744536 | [{"cpu": 0.4931227158526041, "checkpointer": 8}]    | health      |           1 |  1 | test
 2024-09-17 05:57:23.411127 | [{"cpu": 0.9391787279884108, "checkpointer": 21}]   | health      |           1 |  1 | test
 2024-09-17 05:56:55.397735 | [{"cpu": 0.9391787279884108, "checkpointer": 21}]   | health      |           1 |  1 | test
 2024-09-17 05:56:23.33855  | [{"cpu": 0.9391787279884108, "checkpointer": 21}]   | health      |           1 |  1 | test
 2024-09-17 05:55:06.514503 | [{"cpu": 0.9391787279884108, "checkpointer": 21}]   | health      |           1 |  1 | test
 2024-09-17 05:54:19.861201 | [{"cpu": 0.9391787279884108, "checkpointer": 21}]   | health      |           1 |  1 | test
 2024-09-16 10:21:46.461051 | [{"cpu": 0.17688077186140289, "checkpointer": 99}]  | health      |           1 |  1 | test
 2024-09-16 10:20:44.369418 | [{"cpu": 0.015728886965353912, "checkpointer": 22}] | health      |           1 |  1 | test
 2024-09-16 07:33:04.968043 | [{"cpu": 0.1026947434019184, "checkpointer": 0}]    | health      |           1 |  1 | test
 2024-09-16 07:31:22.365576 | [{"cpu": 0.8076059523961426, "checkpointer": 40}]   | health      |           1 |  1 | test
 2024-09-16 07:29:27.054481 | [{"cpu": 0.034741206423261346, "checkpointer": 73}] | health      |           1 |  1 | test
 2024-09-16 07:28:28.229516 | [{"cpu": 0.9956509471402158, "checkpointer": 26}]   | health      |           1 |  1 | test
 2024-09-16 07:27:24.502777 | [{"cpu": 0.8833158321966696, "checkpointer": 40}]   | health      |           1 |  1 | test
 2024-09-16 07:26:39.713865 | [{"cpu": 0.001640287333294862, "checkpointer": 0}]  | health      |           1 |  1 | test
 2024-09-16 07:25:40.452153 | [{"cpu": 0.8729865490332787, "checkpointer": 12}]   | health      |           1 |  1 | test
 2024-09-16 07:24:37.738921 | [{"cpu": 0.8742443008720868, "checkpointer": 62}]   | health      |           1 |  1 | test
 2024-09-16 07:17:10.97184  | [{"cpu": 0.22953988467236924, "checkpointer": 38}]  | health      |           1 |  1 | test
 2024-09-16 07:15:01.259699 | [{"cpu": 1.169480258119128, "checkpointer": 81}]    | health      |           1 |  1 | test
 2024-09-16 07:14:20.124188 | [{"cpu": 1.0811733871405975, "checkpointer": 65}]   | health      |           1 |  1 | test
 2024-09-16 07:13:19.391432 | [{"cpu": 1.3516906697077604, "checkpointer": 38}]   | health      |           1 |  1 | test
 2024-09-16 07:12:55.786994 | [{"cpu": 1.5481504692957349, "checkpointer": 26}]   | health      |           1 |  1 | test
```


Model Output:

```2024/09/18 20:40:28 Title: Metrics Analytics for Healthy Database Performance

Introduction:

In this report, we will analyze the health metrics of a single PostgreSQL database. These metrics, which include CPU usage, checkpointer frequency, and number of open files, are crucial indicators of database performance that can help identify potential issues before they become significant problems. By providing insight into these metrics, we hope to provide recommendations for maintaining optimal database performance.

Metric Name: Health:

1. CPU Usage (0.4931227158526041)
   Metric: CPU usage is the percentage of time that the database server is running on a CPU. A high CPU usage can indicate that the database is being overloaded, leading to slower performance and increased memory usage. To mitigate this issue, we recommend tuning the database configuration settings, such as increasing the number of threads or reducing the maximum connections allowed by the database server.
   Insight: This metric indicates that there are likely more than two CPUs being used, indicating that the database may be experiencing a high workload. To mitigate this issue, it may be beneficial to adjust the number of threads in the database server configuration and/or increase the maximum connections allowed by the server.

2. Checkpointer Frequency (8)
   Metric: The frequency with which the PostgreSQL database checkpointer is run. A low checkpointer frequency indicates that there may be issues with the database or storage, such as data corruption or inefficient indexes. To address this issue, we recommend regularly updating and tuning the database's storage engine, particularly for large datasets or if there are known performance issues associated with in-memory databases.
   Insight: This metric indicates that the database is performing poorly, which may be due to data corruption or an inefficient index configuration. To mitigate this issue, we recommend regularly updating and tuning the database's storage engine.

3. Number of Open Files (21)
   Metric: The number of file descriptors used by the PostgreSQL database server. A high number of open files can indicate that the database is experiencing a memory leak or other issues with the file system, which may lead to slower performance and increased disk I/O. To address this issue, we recommend regularly monitoring and tuning the database's storage and file system configuration settings.
   Insight: This metric indicates that there are likely more than two open files being used by the database server, indicating a memory leak or other issues with the file system. To mitigate this issue, it may be beneficial to regularly monitor and tune the database's storage and file system configuration settings.

Conclusion:

In conclusion, we have analyzed the health metrics of a single PostgreSQL database. We have observed that CPU usage is high, checkpointer frequency is low, and number of open files are high. To address these issues, we recommend regularly monitoring and tuning the database's storage and file system configuration settings, particularly for large datasets or if there are known performance issues associated with in-memory databases. By following these recommendations, we hope to provide optimal database performance for our target application```

**Comment by destrex271** [comment]
This is one more example of what is being generated with the demo database in pgwatch3:

```
I can provide you with a summary of the metrics and observations observed during the checksum failures at pgwatch2. The key findings are:

1. Seed failures: pgwatch2 detected five failed checksums out of 472 total checksums (0.0009%). These were identified as follows:
   - PostgreSQL was running with a single-threaded server process, which caused some transactions to be committed before the other thread could update the checksum table. This led to the failed checksums.
   - The number of concurrent transactions exceeded the threshold configured by postgres (1024) in pgwatch2's configuration file.

2. In_recovery_int: This metric shows whether the server was shutting down gracefully or not. It was observed that some instances were in recovery mode at the time of the failure, which may have caused issues with checksum checkouts from other threads. However, this metric showed a low value (0.5%) during postgres' shutdown process, indicating that these checksums might have been checked out correctly.

3. In_recovery_int: This metric also showed that at least two instances were in recovery mode when the failure occurred. Some checks failed due to lack of resources or other reasons outside of pgwatch2's control, resulting in incomplete checksum checkouts for these instances.

4. Invaliid indexes: pgwatch2 identified 165 invaild indexes out of the 209 total indexes (73%) checked during the checksum failures. This may be due to various reasons, such as the postgres configuration being incorrect or a lack of resources for pgwatch2 to analyze.

5. Invaliid indexes: pgwatch2 identified 165 invaild indexes out of the 209 total indexes (73%) checked during the checksum failures. This may be due to various reasons, such as the postgres configuration being incorrect or a lack of resources for pgwatch2 to analyze.

The observed values do not indicate any significant issues with checksum generation, storage or analysis at pgwatch2 during the checks. However, some instances were observed in recovery mode, which may have resulted in incomplete checksums being checked out by other threads.
```
although it seems to be a little bad with the grammar

**Comment by destrex271** [comment]
> Looks really cool! We need to make it less chatty, IMHO, otherwise really cool!

Sure! will do that

File: cmd/llama_receiver/llama_receiver.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+	if err != nil {
+		log.Println("[ERROR]: unable to setup tables: ", err)
+		return nil, err
+	}
+
+	return recv, nil
+}
+
+func (r *LlamaReceiver) SetupTables() error {
+	_, err := r.DbConn.Exec(r.Ctx, `CREATE TABLE IF NOT EXISTS Db(id serial Primary key, dbname varchar(255))`)
```
Comment (pashagolub) [inline cmd/llama_receiver/llama_receiver.go:91]:
```suggestion
	_, err := r.DbConn.Exec(r.Ctx, `CREATE TABLE IF NOT EXISTS Db(id serial Primary key, dbname text)`)
```
https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default

File: cmd/llama_receiver/llama_receiver.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+	if err != nil {
+		log.Println("[ERROR]: unable to create Db table : " + err.Error())
+		return err
+	}
+
+	_, err = r.DbConn.Exec(r.Ctx, `CREATE TABLE IF NOT EXISTS Measurement (
+		created_time TIMESTAMP NOT NULL DEFAULT(NOW() AT TIME ZONE 'UTC'),
+		data JSONB,
+		metric_name VARCHAR(255),
```
Comment (pashagolub) [inline cmd/llama_receiver/llama_receiver.go:100]:
```suggestion
		metric_name TEXT,
```

Also, we need to use the same capitalization through the whole code. Either types are capitalized everywhere, or lowercased

File: cmd/llama_receiver/llama_receiver.go (hunk 1 above)
Comment (destrex271) [reply cmd/llama_receiver/llama_receiver.go:91]: damn, I was not aware of that 
Thanks!

File: cmd/llama_receiver/llama_receiver.go (hunk 2 above)
Comment (destrex271) [reply cmd/llama_receiver/llama_receiver.go]: okkk, I'll push a separate commit for that

**Comment by destrex271** [comment]
@pashagolub I think this is it for this PR. Would love to get your reviews.

**Review by pashagolub — APPROVED** [review]

File: cmd/llama_receiver/llama_receiver.go [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+	}
+
+	// Convert measurement to json
+	jsonData, err := json.Marshal(msg.Data)
+	if err != nil {
+		return err
+	}
+
+	// insert measurements with current timestamp(default) into table Measurement
+	_, err = conn.Exec(r.Ctx, fmt.Sprintf(`INSERT INTO Measurement(data, database_id, metric_name) VALUES('%s', %d, '%s')`, string(jsonData), id, msg.MetricName))
```
Comment (pashagolub) [inline cmd/llama_receiver/llama_receiver.go:199]: we must use parameters here

**Review by pashagolub — CHANGES_REQUESTED** [review]

File: cmd/llama_receiver/llama_receiver.go (hunk 3 above)
Comment (destrex271) [reply cmd/llama_receiver/llama_receiver.go:199]: okk, will make this changes

**Comment by destrex271** [comment]
@pashagolub fixed the SQL formatting

**Review by pashagolub — APPROVED** [review]

**Comment by destrex271** [comment]
Merging!
